pkg/cloud/services/loadbalancer/loadbalancer.go

```diff
@@ -153,71 +115,190 @@ func (s *Service) ReconcileLoadBalancer(openStackCluster *infrav1.OpenStackClust
+func getAPIServerFloatingIP(openStackCluster *infrav1.OpenStackCluster) (string, error) {
+	switch {
+	// The floating IP was created previously
+	case openStackCluster.Status.APIServerLoadBalancer != nil && openStackCluster.Status.APIServerLoadBalancer.IP != "":
+		return openStackCluster.Status.APIServerLoadBalancer.IP, nil
 
-		pool, err := s.getOrCreatePool(openStackCluster, lbPortObjectsName, listener.ID, lb.ID, lb.Provider)
+	// Explicit floating IP in the cluster spec
+	case openStackCluster.Spec.APIServerFloatingIP != "":
+		return openStackCluster.Spec.APIServerFloatingIP, nil
+
+	// An IP address is specified explicitly in the control plane endpoint
+	case openStackCluster.Spec.ControlPlaneEndpoint.IsValid():
+		floatingIPAddress, err := lookupHost(openStackCluster.Spec.ControlPlaneEndpoint.Host)
 		if err != nil {
-			return false, err
+			return "", fmt.Errorf("lookup host: %w", err)
 		}
+		return floatingIPAddress, nil
+	}
 
-		if err := s.getOrCreateMonitor(openStackCluster, lbPortObjectsName, pool.ID, lb.ID); err != nil {
-			return false, err
+	return "", nil
+}
+
+// getCanonicalAllowedCIDRs gets a filtered list of CIDRs which should be allowed to access the API server loadbalancer.
+// Invalid CIDRs are filtered from the list and emil a warning event.
+// It returns a canonical representation that can be directly compared with other canonicalized lists.
+func getCanonicalAllowedCIDRs(openStackCluster *infrav1.OpenStackCluster) []string {
+	allowedCIDRs := []string{}
+
+	if len(openStackCluster.Spec.APIServerLoadBalancer.AllowedCIDRs) > 0 {
+		allowedCIDRs = append(allowedCIDRs, openStackCluster.Spec.APIServerLoadBalancer.AllowedCIDRs...)
+
+		// In the first reconciliation loop, only the Ready field is set in openStackCluster.Status
+		// All other fields are empty/nil
+		if openStackCluster.Status.Bastion != nil {
+			if openStackCluster.Status.Bastion.FloatingIP != "" {
+				allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Bastion.FloatingIP)
+			}
+
+			if openStackCluster.Status.Bastion.IP != "" {
+				allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Bastion.IP)
+			}
 		}
 
-		if allowedCIDRsSupported {
-			// Skip reconciliation if network status is nil (e.g. during clusterctl move)
-			if openStackCluster.Status.Network != nil {
-				if err := s.getOrUpdateAllowedCIDRS(openStackCluster, listener); err != nil {
-					return false, err
+		if openStackCluster.Status.Network != nil {
+			for _, subnet := range openStackCluster.Status.Network.Subnets {
+				if subnet.CIDR != "" {
+					allowedCIDRs = append(allowedCIDRs, subnet.CIDR)
 				}
-				allowedCIDRs = listener.AllowedCIDRs
+			}
+
+			if openStackCluster.Status.Router != nil && len(openStackCluster.Status.Router.IPs) > 0 {
+				allowedCIDRs = append(allowedCIDRs, openStackCluster.Status.Router.IPs...)
 			}
 		}
 	}
-	lbStatus.AllowedCIDRs = allowedCIDRs
 
-	return false, nil
+	// Filter invalid CIDRs and convert any IPs into CIDRs.
```

Comment:
couldn't we keep using `validateIPs` function?

Reply:
It's a toss-up, but I thought it was too small to be a separate function with only one caller. This function is small enough now, I think it's clearer to inline it here.

Also it doesn't just validate, it also filters, which you wouldn't know just from its name so you'd have to go read the function intentionally to notice that.